express-notes-api/server.js

```diff
@@ -0,0 +1,122 @@
+}
+
+async function writeData(data) {
+  const stringify = JSON.stringify(data, null, 2);
+  await writeFile('data.json', stringify);
+}
+
+app.get('/api/notes', async (req, res) => {
+  try {
+    const data = await readData();
+    const notes = data.notes;
+    const listNotes = [];
+    for (const key in notes) {
+      listNotes.push(notes[key]);
+    }
+    res.status(200).json(listNotes);
+  } catch (error) {
+    console.error(error);
+    res.status(404).json({ error: 'An unexpected error occured' });
+  }
+});
+
+app.get('/api/notes/:id', async (req, res) => {
+  try {
+    const { notes } = await readData();
+    const notesId = Number(req.params.id);
+    if (notesId < 0) {
+      res.status(400).json({ error: 'Note ID must be a positive integer.' });
+      return;
+    }
+    if (!notes[notesId]) {
+      res.status(404).json({ error: `There is no note with that ${notesId}.` });
+      return;
+    }
+    res.status(200).json(notes[notesId]);
+  } catch (error) {
+    console.error(error);
+    res.status(404).json({ error: 'An unexpected error occured.' });
+  }
+});
+
+app.post('/api/notes', async (req, res) => {
+  try {
+    const { content } = req.body;
+    if (content === undefined) {
+      res.status(400).json({ error: 'Content is a required field.' });
+      return;
+    }
+    const data = await readData();
+    const note = {
+      id: data.nextId,
+      content,
+    };
+    data.notes[note.id] = note;
+    data.nextId++;
+    await writeData(data);
+    res.status(201).json(note);
+  } catch (error) {
+    console.error(error);
+    res.status(500).json({ error: 'An unexpected error occured.' });
+  }
+});
+
+app.delete('/api/notes/:id', async (req, res) => {
+  try {
+    const notesId = Number(req.params.id);
+    if (notesId < 0) {
+      res.status(400).json({ error: 'Note ID must be a positive integer.' });
+      return;
+    }
+    const data = await readData();
+    if (data.notes[notesId] === undefined) {
+      res.status(404).json({ error: `There is no note with that ${notesId}.` });
+      return;
+    }
+    delete data.notes[notesId];
+    await writeData(data);
+    res.sendStatus(204);
+  } catch (error) {
+    console.error(error);
+    res.status(500).json({ error: 'An unexpected error occured.' });
+  }
+});
+
+app.put('/api/notes/:id', async (req, res) => {
+  try {
+    const notesId = Number(req.params.id);
+    if (notesId < 0) {
+      res.status(400).json({ error: 'Note ID must be a positive integer.' });
+      return;
+    }
+    const { content } = req.body;
+    if (content === undefined) {
+      res.status(400).json({ error: 'Content is a required field.' });
+      return;
+    }
+    const data = await readData();
+    if (data.notes[notesId] === undefined) {
+      res.status(404).json({ error: `There is no note with that ${notesId}.` });
+      return;
+    }
+    const note = { notesId, content };
+    data.notes[note.id] = note;
```

Comment:
Oops. Above, I said there's no bug now, because I thought that record had been inserted in `post`, but actually, I realize from here, that it was inserted with `put`, because there is a bug here.

At line 110, you create `note` as an object with two fields, one named `notesId` and one named `content`. So here, in line 111, `note.id` is not defined since `note` does not have a field named `id`. The bug is easily fixed, you just need to create `note` in line 110 with an `id` field instead of a `notesId` field.

Reply:
Sounds good. You could also have avoided changing all the `notesId`s by doing this:
`const note = { id: notesId, content };`

Reply:
Thank you! I will remember that for next time haha